File: services/api/internal/sentry/sentry.go
// Package sentry provides error tracking integration with Sentry/GlitchTip.
package sentry

import (
	"context"
	"fmt"
	"time"

	"github.com/getsentry/sentry-go"
	"github.com/irfndi/match-bot/services/api/internal/config"
)

// Init initializes Sentry with the given configuration.
// Returns nil if Sentry is disabled or DSN is empty (graceful degradation).
func Init(cfg config.Config) error {
	if !cfg.EnableSentry || cfg.SentryDSN == "" {
		return nil // Graceful degradation
	}

	err := sentry.Init(sentry.ClientOptions{
		Dsn:         cfg.SentryDSN,
		Environment: cfg.SentryEnvironment,
		Release:     "meetsmatch-api@1.0.0",
		BeforeSend: func(event *sentry.Event, hint *sentry.EventHint) *sentry.Event {
			sanitizeEvent(event)
			return event
		},
	})
	if err != nil {
		return fmt.Errorf("sentry initialization failed: %w", err)
	}

	return nil
}

// Flush flushes any buffered events before shutdown.
func Flush(timeout time.Duration) {
	sentry.Flush(timeout)
}

// CaptureError captures an error with optional context.
func CaptureError(err error, tags map[string]string, extras map[string]interface{}) {
	if err == nil {
		return
	}

	hub := sentry.CurrentHub().Clone()
	scope := hub.Scope()

	for k, v := range tags {
		scope.SetTag(k, v)
	}
	for k, v := range extras {
		scope.SetExtra(k, v)
	}

	hub.CaptureException(err)
}

// CaptureErrorWithContext captures an error with request context.
func CaptureErrorWithContext(ctx context.Context, err error, tags map[string]string, extras map[string]interface{}) {
	if err == nil {
		return
	}

	hub := sentry.GetHubFromContext(ctx)
	if hub == nil {
		hub = sentry.CurrentHub().Clone()
	}

	scope := hub.Scope()

	// Add context-derived info
	if userID, ok := ctx.Value(contextKeyUserID).(string); ok {
		scope.SetUser(sentry.User{ID: userID})
	}
	if requestID, ok := ctx.Value(contextKeyRequestID).(string); ok {
		scope.SetTag("request_id", requestID)
	}

	for k, v := range tags {
		scope.SetTag(k, v)
	}
	for k, v := range extras {
		scope.SetExtra(k, v)
	}

	hub.CaptureException(err)
}

// AddBreadcrumb adds a breadcrumb to the current scope.
func AddBreadcrumb(category, message string, level sentry.Level, data map[string]interface{}) {
	sentry.AddBreadcrumb(&sentry.Breadcrumb{
		Category: category,
		Message:  message,
		Level:    level,
		Data:     data,
	})
}

// Context key types for type-safe context values
type contextKey string

const (
	contextKeyUserID    contextKey = "user_id"
	contextKeyRequestID contextKey = "request_id"
)

// WithUserID returns a new context with the user ID set.
func WithUserID(ctx context.Context, userID string) context.Context {
	return context.WithValue(ctx, contextKeyUserID, userID)
}

// WithRequestID returns a new context with the request ID set.
func WithRequestID(ctx context.Context, requestID string) context.Context {
	return context.WithValue(ctx, contextKeyRequestID, requestID)
}

// sanitizeEvent removes sensitive data from Sentry events.
func sanitizeEvent(event *sentry.Event) {
	if event.Request != nil {
		delete(event.Request.Headers, "Authorization")
		delete(event.Request.Headers, "Cookie")
		delete(event.Request.Headers, "X-Api-Key")
	}
}

File: services/bot/src/lib/sentry.ts
import * as Sentry from '@sentry/node';
import { Effect } from 'effect';

export interface SentryConfig {
  dsn: string;
  environment: string;
  tracesSampleRate: number;
  release: string;
  enabled: boolean;
}

/**
 * Parse and validate a sample rate to ensure it's between 0 and 1.
 */
const parseSampleRate = (value: string | undefined, defaultValue: number): number => {
  const parsed = Number.parseFloat(value || String(defaultValue));
  if (Number.isNaN(parsed)) {
    console.warn(`Invalid sample rate "${value}", using default ${defaultValue}`);
    return defaultValue;
  }
  // Clamp to valid range [0, 1]
  return Math.max(0, Math.min(1, parsed));
};

/**
 * Load Sentry configuration from environment variables.
 * Returns disabled config if SENTRY_DSN is not set (graceful degradation).
 */
export const loadSentryConfig = (): SentryConfig => ({
  dsn: process.env.SENTRY_DSN || '',
  environment: process.env.SENTRY_ENVIRONMENT || 'development',
  tracesSampleRate: parseSampleRate(process.env.SENTRY_TRACES_SAMPLE_RATE, 0.2),
  release: process.env.SENTRY_RELEASE || 'meetsmatch-bot@dev',
  enabled: process.env.ENABLE_SENTRY === 'true' && !!process.env.SENTRY_DSN,
});

/**
 * Initialize Sentry with the given configuration.
 * Does nothing if disabled (graceful degradation).
 */
export const initSentry = (config: SentryConfig): void => {
  if (!config.enabled) {
    console.log('Sentry is disabled or SENTRY_DSN not set');
    return;
  }

  Sentry.init({
    dsn: config.dsn,
    environment: config.environment,
    tracesSampleRate: config.tracesSampleRate,
    release: config.release,
    beforeSend(event) {
      // Sanitize sensitive data
      if (event.request?.headers) {
        delete event.request.headers.authorization;
        delete event.request.headers.cookie;
      }
      return event;
    },
  });

  console.log(`Sentry initialized for environment: ${config.environment}`);
};

/**
 * Capture an error with optional context.
 */
export const captureError = (
  error: unknown,
  context?: {
    tags?: Record<string, string>;
    extras?: Record<string, unknown>;
    userId?: string;
  },
): void => {
  if (!error) return;

  Sentry.withScope((scope) => {
    if (context?.tags) {
      for (const [key, value] of Object.entries(context.tags)) {
        scope.setTag(key, value);
      }
    }
    if (context?.extras) {
      for (const [key, value] of Object.entries(context.extras)) {
        scope.setExtra(key, value);
      }
    }
    if (context?.userId) {
      scope.setUser({ id: context.userId });
    }

    if (error instanceof Error) {
      Sentry.captureException(error);
    } else {
      // Wrap non-Error types in an Error to preserve call stack
      const wrappedError = new Error(String(error));
      wrappedError.name = 'NonErrorException';
      Sentry.captureException(wrappedError);
    }
  });
};

/**
 * Add a breadcrumb for debugging.
 */
export const addBreadcrumb = (
  category: string,
  message: string,
  level: 'debug' | 'info' | 'warning' | 'error' = 'info',
  data?: Record<string, unknown>,
): void => {
  Sentry.addBreadcrumb({
    category,
    message,
    level,
    data,
  });
};

/**
 * Flush pending events (call before shutdown).
 */
export const flushSentry = async (timeout = 2000): Promise<void> => {
  await Sentry.flush(timeout);
};

/**
 * Effect-based error capture that logs and reports to Sentry.
 * Replaces console.error in Effect.catchAll handlers.
 */
export const captureEffectError =
  <E>(context: string, userId?: string) =>
  (error: E): Effect.Effect<void, never> =>
    Effect.sync(() => {
      console.error(`${context}:`, error);
      captureError(error, {
        tags: { context },
        userId,
      });
    });
